### PR TITLE
Dependency resolution details / results are exposed via build operations

### PR DESCRIPTION
**Pull request by breskeby**
This PR adds Details/Result types to the BuildOperation wrapping dependency resolution. The build scan plugin will use this interface to move away from directly accessing configurations by iterating over `project.allProjects`.

I raise this PR to get early feedback while working in parallel on adding more integration test coverage for this.

### CODE REVIEW

**Comment by breskeby** [comment]
related CI build: https://builds.gradle.org/viewLog.html?buildTypeId=Gradle_Check_Stage_BranchBuildAccept_Trigger&buildId=9185298&branch_Gradle_Check_Stage_BranchBuildAccept_Trigger=breskeby/resolve-deps-build-operations

**Comment by breskeby** [comment]
crucible review is available at https://code-review.gradle.org/cru/REVIEW-6570

File: subprojects/core/src/main/java/org/gradle/internal/operations/trace/SerializedOperationFinish.java [hunk 1]
```diff
@@ -37,11 +39,27 @@
     SerializedOperationFinish(BuildOperationDescriptor descriptor, OperationFinishEvent finishEvent) {
         this.id = ((OperationIdentifier) descriptor.getId()).getId();
         this.endTime = finishEvent.getEndTime();
-        this.result = finishEvent.getResult();
+        this.result = transform(finishEvent.getResult());
         this.resultClassName = result == null ? null : finishEvent.getResult().getClass().getName();
         this.failureMsg = finishEvent.getFailure() == null ? null : finishEvent.getFailure().toString();
     }
 
+    private Object transform(Object details) {
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/internal/operations/trace/SerializedOperationFinish.java]: If we want to keep going this path (serializing the details), we should make the serializer/deserializer pluggable and put it in the right subproject. I'm not sure why we try to serialize all details objects though. @ldaley what is your take on this?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 2]
```diff
@@ -480,11 +482,24 @@ public void run(BuildOperationContext context) {
...
+                        return incoming.getResolutionResult().getRoot();
+                    }
+                });
             }
 
             @Override
             public BuildOperationDescriptor.Builder description() {
-                return BuildOperationDescriptor.displayName("Resolve dependencies of " + identityPath).progressDisplayName("Resolve dependencies " + identityPath);
+                return BuildOperationDescriptor.displayName("Resolve dependencies of " + identityPath)
+                    .progressDisplayName("Resolve dependencies " + identityPath)
```
Comment (oehme) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: This is not new, but bothered me before: Missing "of"

File: subprojects/core/src/main/java/org/gradle/internal/operations/trace/SerializedOperationFinish.java (hunk 1 above)
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/internal/operations/trace/SerializedOperationFinish.java]: we currently only use it for our `-Dorg.gradle.internal.operations.trace` flag to get the build operations to disk. This is also used in the `BuildOperationTestFixture`

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java (hunk 2 above)
Comment (breskeby) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: fixed this.

**Comment by breskeby** [comment]
Added test coverage for normal builds, composite builds and detached configurations. When this passed CI, I plan to merge this into master if there are no objections as this is blocking build scan plugin work.

File: subprojects/core/src/main/java/org/gradle/internal/operations/trace/SerializedOperationFinish.java (hunk 1 above)
Comment (ldaley) [reply subprojects/core/src/main/java/org/gradle/internal/operations/trace/SerializedOperationFinish.java]: @oehme as Rene says, this is just for testing and debugging. 

I think we are close to the threshold of needing a better solution, but we aren't there yet.

**Pull request merged**